### PR TITLE
[testing] Share Persist state across SLT runs

### LINKED ISSUE #20677: test/sqllogictest/id_reuse.slt is flaky: initialized shard should have at least one diff
### Buildkite link

https://buildkite.com/materialize/tests/builds/59902#018971e5-d349-4d3f-bd8f-9bd276acc7f8

### Relevant log output

```source
rliest live diff to advance: v129 vs v129
[2023-07-20T06:24:54Z] 2023-07-20T06:24:54.211959Z  WARN mz_persist_client::internal::state_versions: logic error: fetch_current_state refetch expects earliest live diff to advance: v129 vs v129
[2023-07-20T06:24:54Z] 2023-07-20T06:24:54.213085Z  WARN mz_persist_client::internal::state_versions: logic error: fetch_current_state refetch expects earliest live diff to advance: v129 vs v129
[2023-07-20T06:24:54Z] 2023-07-20T06:24:54.213872Z  WARN mz_persist_client::internal::state_versions: logic error: fetch_current_state refetch expects earliest live diff to advance: v129 vs v129
[2023-07-20T06:24:54Z] 2023-07-20T06:24:54.214516Z  WARN mz_persist_client::internal::state_versions: logic error: fetch_current_state refetch expects earliest live diff to advance: v129 vs v129
[2023-07-20T06:24:54Z] thread 'tokio-runtime-worker' panicked at 'initialized shard should have at least one diff', /var/lib/buildkite-agent/builds/buildkite-builders-d43b1b5-i-0b64900c0c389b5f3-1/materialize/tests/src/persist-client/src/internal/state_versions.rs:381:26
[2023-07-20T06:24:54Z] stack backtrace:
[2023-07-20T06:24:54Z]    0: rust_begin_unwind
[2023-07-20T06:24:54Z]    1: core::panicking::panic_fmt
[2023-07-20T06:24:54Z]    2: core::panicking::panic_display
[2023-07-20T06:24:54Z]    3: core::panicking::panic_str
[2023-07-20T06:24:54Z]    4: core::option::expect_failed
[2023-07-20T06:24:54Z]    5: <mz_persist_client::internal::state_versions::StateVersions>::fetch_current_state::<mz_repr::timestamp::Timestamp>::{closure#0}
[2023-07-20T06:24:54Z]    6: <mz_persist_client::internal::gc::GarbageCollector<mz_storage_client::types::sources::SourceData, (), mz_repr::timestamp::Timestamp, i64>>::gc_and_truncate::{closure#0}
[2023-07-20T06:24:54Z]    7: <tracing::instrument::Instrumented<<mz_persist_client::internal::gc::GarbageCollector<mz_storage_client::types::sources::SourceData, (), mz_repr::timestamp::Timestamp, i64>>::gc_and_truncate::{closure#0}> as core::future::future::Future>::poll
[2023-07-20T06:24:54Z]    8: <mz_persist_client::internal::gc::GarbageCollector<mz_storage_client::types::sources::SourceData, (), mz_repr::timestamp::Timestamp, i64>>::new::{closure#1}::{closure#1}
[2023-07-20T06:24:54Z]    9: <tokio::loom::std::unsafe_cell::UnsafeCell<tokio::runtime::task::core::Stage<<mz_persist_client::internal::gc::GarbageCollector<mz_storage_client::types::sources::SourceData, (), mz_repr::timestamp::Timestamp, i64>>::new::{closure#1}::{closure#1}>>>::with_mut::<core::task::poll::Poll<(mz_persist_client::internal::maintenance::RoutineMaintenance, mz_persist_client::internal::gc::GcResults)>, <tokio::runtime::task::core::Core<<mz_persist_client::internal::gc::GarbageCollector<mz_storage_client::types::sources::SourceData, (), mz_repr::timestamp::Timestamp, i64>>::new::{closure#1}::{closure#1}, alloc::sync::Arc<tokio::runtime::scheduler::multi_thread::handle::Handle>>>::poll::{closure#0}>
[2023-07-20T06:24:54Z]   10: <tokio::runtime::task::core::Core<<mz_persist_client::internal::gc::GarbageCollector<mz_storage_client::types::sources::SourceData, (), mz_repr::timestamp::Timestamp, i64>>::new::{closure#1}::{closure#1}, alloc::sync::Arc<tokio::runtime::scheduler::multi_thread::handle::Handle>>>::poll
[2023-07-20T06:24:54Z]   11: <tokio::runtime::task::harness::Harness<<mz_persist_client::internal::gc::GarbageCollector<mz_storage_client::types::sources::SourceData, (), mz_repr::timestamp::Timestamp, i64>>::new::{closure#1}::{closure#1}, alloc::sync::Arc<tokio::runtime::scheduler::multi_thread::handle::Handle>>>::poll
[2023-07-20T06:24:54Z]   12: <tokio::runtime::scheduler::multi_thread::worker::Context>::run_task
[2023-07-20T06:24:54Z]   13: <tokio::runtime::scheduler::multi_thread::worker::Context>::run
[2023-07-20T06:24:54Z]   14: <tokio::macros::scoped_tls::ScopedKey<tokio::runtime::scheduler::multi_thread::worker::Context>>::set::<tokio::runtime::scheduler::multi_thread::worker::run::{closure#0}, ()>
[2023-07-20T06:24:54Z]   15: tokio::runtime::scheduler::multi_thread::worker::run
[2023-07-20T06:24:54Z]   16: <tokio::loom::std::unsafe_cell::UnsafeCell<tokio::runtime::task::core::Stage<tokio::runtime::blocking::task::BlockingTask<<tokio::runtime::scheduler::multi_thread::worker::Launch>::launch::{closure#0}>>>>::with_mut::<core::task::poll::Poll<()>, <tokio::runtime::task::core::Core<tokio::runtime::blocking::task::BlockingTask<<tokio::runtime::scheduler::multi_thread::worker::Launch>::launch::{closure#0}>, tokio::runtime::blocking::schedule::BlockingSchedule>>::poll::{closure#0}>
[2023-07-20T06:24:54Z]   17: <tokio::runtime::task::core::Core<tokio::runtime::blocking::task::BlockingTask<<tokio::runtime::scheduler::multi_thread::worker::Launch>::launch::{closure#0}>, tokio::runtime::blocking::schedule::BlockingSchedule>>::poll
[2023-07-20T06:24:54Z]   18: <tokio::runtime::task::harness::Harness<tokio::runtime::blocking::task::BlockingTask<<tokio::runtime::scheduler::multi_thread::worker::Launch>::launch::{closure#0}>, tokio::runtime::blocking::schedule::BlockingSchedule>>::poll
[2023-07-20T06:24:54Z]   19: <tokio::runtime::blocking::pool::Inner>::run
[2023-07-20T06:24:54Z] note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.
[2023-07-20T06:24:54Z] ci/test/slt-fast.sh: line 370:   334 Aborted                 (core dumped) sqllogictest -v "${tests_without_views[@]}" "$@"
[2023-07-20T06:24:55Z] ==> Uploading report for suite 'sqllogictest' to Buildkite Test Analytics
[2023-07-20T06:24:55Z] 422 {
[2023-07-20T06:24:55Z]   "message": "`data` is required"
[2023-07-20T06:24:55Z] }
[2023-07-20T06:24:55Z]
[2023-07-20T06:24:55Z] mzcompose: test case workflow-default failed: running docker compose failed (exit status 134)
[2023-07-20T06:24:55Z] ==> Uploading report for suite 'mzcompose' to Buildkite Test Analytics
[2023-07-20T06:24:55Z] 202 {"id":"3b86fdd9-0511-4649-85b3-7a33f8ca2cc6","run_id":"c3dddc38-4ff3-4b87-9321-53539bf5710f","queued":1,"skipped":0,"errors":[],"run_url":"https://buildkite.com/organizations/materialize/analytics/suites/mzcompose/runs/c3dddc38-4ff3-4b87-9321-53539bf5710f"}
[2023-07-20T06:24:55Z] mzcompose: error: at least one test case failed
```

ci-regexp: fetch_all_live_states refetch expects earliest live diff to advance

### Additional thoughts

_No response_

### PR DESCRIPTION
**Pull request by bkirwi**
### Motivation

Fixes #20677, such as it is. 

### Tips for reviewer

See the linked issue for context, and the code comment for why this change is considered safe.

### Checklist

- [ ] This PR has adequate test coverage / QA involvement has been duly considered.
- [ ] This PR has an associated up-to-date [design doc](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/README.md), is a design doc ([template](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/00000000_template.md)), or is sufficiently small to not require a design.
  
- [ ] If this PR evolves [an existing `$T ⇔ Proto$T` mapping](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/command-and-response-binary-encoding.md) (possibly in a backwards-incompatible way), then it is tagged with a `T-proto` label.
- [ ] If this PR will require changes to cloud orchestration, there is a companion cloud PR to account for those changes that is tagged with the release-blocker label ([example](https://github.com/MaterializeInc/cloud/pull/5021)).
  
- [ ] This PR includes the following [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note):
  -

### CODE REVIEW

**Review by danhhz — APPROVED** [review]

**Comment by bkirwi** [comment]
For the record: https://github.com/MaterializeInc/materialize/pull/20923 was a previous attempt at solving the same SLT issue. After some internal discussion we decided that Persist was "in spec" for this behaviour, so we preferred to fix it by changing how SLT was holding it instead of adding the hack to Persist itself. (We may want to support graceful shutdown in the future, but if so we'd aim to do something that was more intentionally designed than bodged on...)

**Comment by bkirwi** [comment]
@MaterializeInc/surfaces (as the owner) - I'm going to go ahead and merge now that I have one review, since this hopefully helps with an ongoing breakage in CI. I'm happy to take any review comments in a followup PR, though!

**Comment by pH14** [comment]
Retro LGTM! I like this approach 👍

File: src/sqllogictest/src/runner.rs [hunk 1]
```diff
@@ -1695,6 +1697,11 @@ pub struct RunConfig<'a> {
     pub enable_table_keys: bool,
     pub orchestrator_process_wrapper: Option<String>,
     pub system_parameter_defaults: BTreeMap<String, String>,
+    /// Persist state is handled specially because:
+    /// - Persist background workers do not necessarily shut down immediately once the server is
+    ///   shut down, and may panic if their storage is delete out from under them.
```
Comment (antiguru) [inline src/sqllogictest/src/runner.rs:1702]: Nit: spelling (`is delete`) :)

**Pull request merged**